Preserve content edits from automatic update

bwhicks opened the pull request:
This PR addresses issue #183 and allows content editors edits to be preserved from Hathitrust (or other) updates by setting a `protected_fields` flags field. 


- sets flags for any updated fields in those included in basic metadata that are user editable (i.e. collections are not part of this) when an item is saved in the admin
-  displays those flags as non-editable, human readable text in the change view. 
- prevents the `populate_from_bibdata` method of `DigitizedWork` from updating those fields, when they have been flagged.

The implementation makes `protected_fields` unconditionally return an instance of `ProtectedFlags` and explicitly makes the field non-nullable, which avoids `None` handling in custom field methods. It also allows the default to always be `ProtectedFlags.no_flags` (i.e. `0`).

rlskoeser commented on ppa/archive/admin.py:
```diff
@@ -59,6 +60,30 @@ def source_link(self, obj):
+        '''Note any fields in the protected list that have been changed
+        and preserve in database.'''
+        # new object, created from scratch, nothing to track and preserve
+        # NOTE: Is this logic correct? Or should all the fields be preserved?
```
This logic seems right to me - we don't expect (or support) new records for HathiTrust items to be created from django admin.

However, it does occur to me that the protected field flags are hathi specific - I think this logic should only trigger when source=hathi.

bwhicks replied:
I will add a flag for this re: hathi

rlskoeser commented on ppa/archive/admin.py:
```diff
@@ -59,6 +60,30 @@ def source_link(self, obj):
+        if not change:
+            super().save_model(request, obj, form, change)
+            return
+        db_obj = DigitizedWork.objects.get(pk=obj.pk)
```
`DigitizedWork` extends `TrackChangesModel` - is there any reason we can't use the `has_changed` method here?

bwhicks replied:
I think I have a way to use this using `refresh_from_db`; the wrinkle as discussed being the obj is a new creation and not aware of the database. If that doesn't work, I'll note the problem.

bwhicks replied:
No dice on `has_changed` and `refresh_from_db`. In the end, cleanest way is to compare against a DB copy. I've added a note explaining why I'm not using `has_changed`.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -121,6 +122,76 @@ def stats():
         return stats
 
 
+class ProtectedFlags(Flags):
```
These flags are very much specific to `DigitizedWork`, so I think the name should indicate that more clearly. I'm not sure how to do that without making it super long, though. `ProtectedWorkFields` maybe? Or are you using flags here to distinguish from the model field? `ProtectedWorkFieldFlags` ?

bwhicks replied:
That was the use of `Flags` (distinguishing from the model field). I like the two proposed options.

bwhicks replied:
Actually, I like `ProtectedWorkFieldFlags` and `ProtectedWorkField` because it's a `models.Field` subclass.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -184,6 +255,13 @@ class DigitizedWork(TrackChangesModel, Indexable):
+    protected_fields = ProtectedFlagsField(
+        default=ProtectedFlags,
+        help_text='Fields protected from bulk update because they have been '
+                  'manually edited.'
```
I might make this even more specific - not just manually edited, but edited in django admin.

Should probably also mention that it's Hathi-specific.

bwhicks replied:
Agreed. I've made it point more specifically to Hathi.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -283,13 +361,16 @@ def populate_from_bibdata(self, bibdata):
         '''
         # store hathi record id
         self.record_id = bibdata.record_id
+        print(bool(bibdata.marcxml))
```
looks like a leftover from debugging or testing something

bwhicks replied:
Thanks for the spot.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -283,13 +361,16 @@ def populate_from_bibdata(self, bibdata):
             # - clean title: strip trailing space & slash and initial bracket
-            self.title = bibdata.marcxml['245']['a'].rstrip(' /') \
-                .lstrip('[')
+            if not self.protected_fields.title:
```
I wonder if it would be cleaner to handle all the protected field logic in one place? Something like this: collect all the new bibliographic data in a dictionary, and then iterate over the dictionary data and set it on `self` unless the field is flagged as protected. 

I think it would make the code more readable and more efficient - handle the marc logic separate from the protected field logic.

In fact, it might be useful for the protected field logic to be in a separate method - that would make it possible to honor the protected flags if we do any other kind of bulk updates from scripts or console.

rlskoeser replied:
I checked the code climate report, and it's complaining about the repeated/similar code where you're checking for protected fields, so refactoring would resolve that also.

bwhicks replied:
This worked out well. I was able to abstract the check out to a nice, short loop.

rlskoeser commented on ppa/archive/tests/test_models.py:
```diff
@@ -15,13 +15,51 @@
+        # should be a comma-joined list of the verbose names of the
+        # fields, in alphabetical order; avoided hard coding in case
+        # verbose name changes
+        assert str(fields) == ', '.join(comparison)
```
I find this test hard to read; I wonder if the verbose name logic should be split out so it's testable directly?

bwhicks replied:
Rendered a non-issue by not needing the verbose fields

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -121,6 +122,76 @@ def stats():
+        verbose_names = []
+        for field in protected_fields:
+            field = DigitizedWork._meta.get_field(field)
+            # if the field has no actual verbose name, just capitalize
```
There isn't an existing method we can use here?

bwhicks replied:
Sadly, no. The `_meta` is how you get access to field names and they are pretty low-level. Thankfully no no longer an issue.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -121,6 +122,76 @@ def stats():
+        return [field for field in all_fields
+                if getattr(self, field)]
+
+    def __str__(self):
```
Can't you make use of the `to_simple_str` method on flags objects? Maybe just change `|` to `,`? I'm not sure how important it is to change internal name to verbose names, this is a backend interface. Maybe we can get feedback from the project team on that. 

If we do want verbose names, then should probably either iterate over set flags or maybe use simple str as the starting point.

rlskoeser replied:
Also, if we do decide we want verbose names - can we calculate them once and store rather than converting every time `str` is called?

bwhicks replied:
Thanks for this re: `to_simple_str`.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -121,6 +122,76 @@ def stats():
+    def all_fields(cls):
+        return list(ProtectedFlags.__members__.keys())
+
+    def as_list(self):
```
Can't you just use the existing flag `__iter__` method?

bwhicks replied:
Likewise for this! Much simpler even if I have to do a single dot split (shouldn't be too punishing).

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -121,6 +122,76 @@ def stats():
+        return ('ppa.archive.models.ProtectedFlags', ['no_flags'], {})
+
+    @classmethod
+    def all_fields(cls):
```
I thought you should just be able to iterate over the class to get the defined fields, but I see from experimentation that it returns the verbose name from the string method! I'm not sure that's actually the behavior we want?

bwhicks replied:
I think we're done with since we're not doing verbose names.

bwhicks commented:
@rlskoeser I believe this is ready for another pass. I refactored the protected_fields logic so as to take a dictionary of fields and values to set, simplified the logic for the flags significantly, and revised the migrations with appropriate class names, etc.

rlskoeser commented on ppa/archive/models.py:
```diff
@@ -136,50 +136,28 @@ class ProtectedFlags(Flags):
-    def all_fields(cls):
-        return list(ProtectedFlags.__members__.keys())
+    def __str__(self):
+        return ', '.join(sorted(self.to_simple_str().split('|')))
```
I think I would have just done a string replace here, rather than splitting and sorting, since it looks like the flags are returned in the order they are defined on the flag subclass (which is a logical order, since you defined it). But this is fine.

rlskoeser commented:
Updates look good to me.
